versioned/storage/cleanup/src/main/java/org/projectnessie/versioned/storage/cleanup/CutHistoryImpl.java

```diff
@@ -0,0 +1,127 @@
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.storage.cleanup;
+
+import static org.projectnessie.versioned.storage.cleanup.ImmutableCutHistoryScanResult.Builder;
+import static org.projectnessie.versioned.storage.common.logic.Logics.commitLogic;
+import static org.projectnessie.versioned.storage.common.objtypes.StandardObjType.COMMIT;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.function.IntFunction;
+import org.projectnessie.versioned.storage.common.logic.CommitLogic;
+import org.projectnessie.versioned.storage.common.objtypes.CommitObj;
+import org.projectnessie.versioned.storage.common.persist.CloseableIterator;
+import org.projectnessie.versioned.storage.common.persist.Obj;
+import org.projectnessie.versioned.storage.common.persist.ObjId;
+import org.projectnessie.versioned.storage.common.persist.Persist;
+
+public class CutHistoryImpl implements CutHistory {
+  private final CutHistoryParams context;
+  private final RateLimit scanRateLimiter;
+
+  public CutHistoryImpl(CutHistoryParams context, IntFunction<RateLimit> rateLimitIntFunction) {
+    this.context = context;
+    this.scanRateLimiter = rateLimitIntFunction.apply(context.scanObjRatePerSecond());
+  }
+
+  @Override
+  public CutHistoryScanResult identifyAffectedCommits() {
+    Builder result = CutHistoryScanResult.builder();
+
+    Persist persist = context.persist();
+    ObjId cutPoint = context.cutPoint();
+
+    long numObjects = 0;
+    // Identify commits, whose "tails" overlap the cut point.
+    // scanAllCommitLogEntries() returns all commits in no specific order, parents may be scanned
+    // before or after their children.
+    try (CloseableIterator<Obj> scan = persist.scanAllObjects(Collections.singleton(COMMIT))) {
+      while (scan.hasNext()) {
+        scanRateLimiter.acquire();
+
+        CommitObj commit = (CommitObj) scan.next();
+        numObjects++;
+        if (commit.tail().contains(cutPoint)) {
+          result.addAffectedCommitId(commit.id());
+        }
+      }
+
+      result.numScannedObjs(numObjects);
+      return result.build();
+    }
+  }
+
+  @Override
+  public UpdateParentsResult rewriteParents(CutHistoryScanResult scanResult) {
+    ImmutableUpdateParentsResult.Builder result = UpdateParentsResult.builder();
+    CommitLogic commitLogic = commitLogic(context.persist());
+    for (ObjId id : scanResult.affectedCommitIds()) {
+      try {
+        CommitObj commitObj = commitLogic.fetchCommit(id);
+        if (commitObj == null) {
+          result.putFailure(id, new IllegalAccessException("Commit not found: " + id));
+          continue;
+        }
+
+        if (commitObj.tail().isEmpty()) {
+          continue;
+        }
+
+        CommitObj updated =
+            CommitObj.commitBuilder()
+                .from(commitObj)
+                .tail(commitObj.tail().subList(0, 1)) // one direct parent for simplicity
```

Comment:
I suspect this is wrong. The identify-function above does a `.contains` - so I think this must cut at the same index, no?

Reply:
Like the comment says, one parent is preserved for simplicity, but I will add change this to the maximum sublist as an optimization.

Reply:
updated